Supernova bugfixes/improvements

Spacechild1 opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->
Some people have tried to build and use my VSTPlugin extension with Supernova but without success. It turned out that there are several bugs and missing features in Supernova, which make it impossible for VSTPlugin to run correctly. This PR attempts to fix the following issues:

* registering and running unit commands is completely broken
* `mParentIndex` in `Unit` is never set (it's always 0)
* `DoAsynchronousCommand` crashes if `cmdName` is `NULL`. OTOH, scsynth performs a `NULL` check, so you can effectively suppress sending a "/done" message.
* the plugin API methods `SendMsgFromRT` and `SendMsgToRT` are not implemented in Supernova. If a plugin tries to call these methods, it will crash.

EDIT: also fixes the following issues:
* `DoAsynchronousCommand` should stop if a stage returns false (like scsynth)
* missing thread synchronization for system_callback memory pool in certain situations

After applying my fixes / additions, Supernova shows the same behavior as scsynth and VSTPlugin works smoothly. It would be great if this PR could be considered for the next bug fix release!

Thanks, Christof

## Types of changes

<!-- Delete lines that don't apply -->

- Bug fix
- New feature

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] This PR is ready for review

Spacechild1 commented:
Also, I noticed that the logic in `handle_asynchronous_command` in Supernova is pretty different from `PerformAsynchronousCommand` in Scsynth: 
* in Supernova, all stages are called, while in Scsynth subsequent stages are only called if the previous stage function was either `NULL` or returned `true` - expect for the cleanup function, which is always called.
* in Supernova, the /done message is always sent, while in Scsynth it is only sent if stage4 is reached and did not return `false`

is this something I should fix as well or should we just leave it as it is?

sonoro1234 commented:
@Spacechild1 I could find other bad behaviours in supernova as reported in #4400
Nobody is giving feedback, but may be nobody is self-confident about supernova. It seems you digged a little into supernova so may be you could have something to say about the issue and proposed PR #4410 
Thanks

Spacechild1 commented:
oh, that sounds like a nasty race condition bug... I only digged into Supernova so people can use by VSTPlugin UGen. Honestly, I'm not a really much of a SuperCollider user, so fixing bugs in Supernova isn't really on the top of my list :-). Maybe I can have a look during summer, but I can't promise anything...

Spacechild1 commented:
@brianlheim who is actually maintaining Supernova right now?

brianlheim commented:
This PR has merge conflicts now, can you please rebase it according to the instructions [here](https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions#i-made-a-pr-before-the-reformat-and-it-has-merge-conflicts-now-how-do-i-fix-it)? Let me know if you have any questions. Thanks!

brianlheim commented:
> who is actually maintaining Supernova right now?

since supernova is a part of supercollider, everyone who maintains this project also maintains supernova.

Spacechild1 commented:
> 
> 
> > who is actually maintaining Supernova right now?
> 
> since supernova is a part of supercollider, everyone who maintains this project also maintains supernova.

I guess what I wanted to know is who is actually able to review PRs for supernova (because not everyone is familiar with the codebase). But this has actually been answered in the last dev meetings :-). Will rebase soon.

sonoro1234 commented:
> is this something I should fix as well or should we just leave it as it is?

I think that the most general rule is: supernova should mimic scsynth from the client point of view, so I would fix it.

You have another proposal about having only one NRT thread that perhaps could be done in another PR after all this is merged?

Spacechild1 commented:
> so I would fix it.

Alright, will do it!

> You have another proposal about having only one NRT thread that perhaps could be done in another PR after all this is merged?

Can do this as well. But I need you as a tester :-)

sonoro1234 commented:
> Can do this as well. But I need you as a tester :-)

I will be glad to be your tester ;-)

Spacechild1 commented:
@snappizz rebased on top of develop and added some more commits. The PR is now ready for review.

The changes work fine for me so far (tested with VSTPlugin) but it needs a thorough review. Don't hesitate to comment if something seems odd or wrong! Also for @sonoro1234

sonoro1234 commented:
@Spacechild1 
This one is difficult to test. No other plugin has never before used this funcionality, so I guess that it can only  be tested with VSTPlugin.
I will try to get information about how to test it.

Spacechild1 commented:
> No other plugin has never before used this funcionality, so I guess that it can only be tested with VSTPlugin.

You can try the `pluginCmdDemo` command (see UIUGens.cpp). But most of this PR just tries to make sure that the behavior of Supernova matches that of Scsynth and this can be checked by comparing the code. Anyway, with these fixes VSTPlugin seems to work just fine and as you said, I seem to be the only (?) one using this part of the plugin API :-D

sonoro1234 commented:
Do you mean that using
https://github.com/supercollider/supercollider/blob/develop/server/plugins/UIUGens.cpp#L372
is enough?
I thought that you were using additional functionallity

Spacechild1 commented:
> This commands worked correctly before applying this PR!!

Yes, it's just to check that I didn't break anything. See the PR description for what I had to fix/add to make VSTPlugin work on Supernova.

sonoro1234 commented:
With your PR instead of done first b_info last, it behaves as scsynth b_info first done last.

Spacechild1 commented:
Ah yes, i've fixed stage 4 according to https://github.com/supercollider/supercollider/pull/4460

sonoro1234 commented:
I am trying VSTPlugin. 
scsynth was tried for search function only.
Managed to build it for supernova also but I am getting a crash in supernova but also in scsynth.
Posted an issue: https://git.iem.at/pd/vstplugin/issues/26

Spacechild1 commented:
FYI: the bug in https://git.iem.at/pd/vstplugin/issues/26 turned out be a mingw32 bug, not related to this PR.

@sonoro1234 would you approve this PR?

sonoro1234 commented:
@snappizz  I think that your requested changes had been applied already!!

Spacechild1 commented:
@snappizz Would you approve this PR? Your requested changes (i.e. reformatting) have been applied.

snappizz commented:
thanks for the rebase, i have dismissed my requested changes.

unfortunately, to focus my efforts on more important parts of SC, i can't assist with supernova maintenance anymore. sorry :(

sonoro1234 commented:
@brianlheim 
Can I merge this?

brianlheim commented:
no

brianlheim commented:
@Spacechild1 this PR now has merge conflicts, can you please rebase it and resolve merge conflicts?

Spacechild1 commented:
Rebasing somehow was too complicated, so I gave up and did a merge. Is this a problem?

brianlheim commented:
@Spacechild1 yes, it's much cleaner if you rebase. Perhaps if you squash this branch first? Then you don't need to resolve 14 rounds of merge conflicts.

sonoro1234 commented:
@brianlheim @Spacechild1 
Just to understand the github process:
You are saying that there are merge conflicts but I cant see them in the `Merge pull request` box.
Where should I have looked?

Spacechild1 commented:
@brianlheim in the end, "rebase" turned out to be simple. The problem was that the first 7 commits were still in the old formatting. After resolving the first (big) conflict, the remaining commits applied cleanly.

Spacechild1 commented:
> You are saying that there are merge conflicts but I cant see them in the Merge pull request box.

@sonoro1234 at the time you've looked at the PR, I've already *merged* develop into my branch and fixed the conflict. In the meantime I've managed to rebase and force pushed (so you the merge commit has disappeared).

Spacechild1 commented:
@brianlheim just to know: are planning to merge this or is this postponed for the next release? I'm asking because of the notes from the SC dev meetings:

> Although there are some high-priority tasks remaining, we decided against adding ANY new things to 3.10.3 largely for psychological reasons --

Just want to say that this PR a) doesn't touch existing behaviour (because things were either completely broken or not implemented) and b) it has been successfully tested by Victor and me with VSTPlugin.

brianlheim commented:
@Spacechild1 if it's merged before the 3.10.3 beta then it can go in. The point of that note (as I understand the conversation, others can correct me) was that the 3.10.3 release should not depend on merging any additional PRs.

Spacechild1 commented:
> if it's merged before the 3.10.3 beta then it can go in.

I guess this is up to you?

brianlheim commented:
> I guess this is up to you?

no; any experienced reviewer is free to review this PR and merge it. we're currently trying a new project org plan where new reviewers can get experience by reviewing PRs and having their reviews 'reviewed' by a more experienced maintainer before merge. @sonoro1234 is part of that. if any of this is unclear let's please move discussion elsewhere so this PR thread stays on topic.

Spacechild1 commented:
@sonoro1234 already approved this PR, that's why I was asking! 

or do you mean a new reviewer should have a look (good luck to them :-) and Victor should supervise? 

> if any of this is unclear let's please move discussion elsewhere so this PR thread stays on topic.

I don't have time to switch to another communication channel right now, feel free to delete our conversation later if it's too off-topic.

sonoro1234 commented:
I must be dumb, all this seems kafkian to me.

brianlheim commented:
@Spacechild1 and I have moved this convo to private email thread :)

sonoro1234 commented:
My purpose was to integrate all server commits in one personal branch in order to being able to compile servers with changes not added to develop

The first step was checking-out to Spacechild1/supernova_fixes
Trying to pull from Spacechild1/supernova_fixes (from which I dont have last commit) gives me a merge conflict
```
CONFLICT (content): Merge conflict in server/supernova/sc/sc_osc_handler.cpp
Automatic merge failed; fix conflicts and then commit the result.

git did not exit cleanly (exit code 1) (3416 ms @ 30/09/2019 12:26:49)

```

Spacechild1 commented:
> from which I dont have last commit

@sonoro1234 that last commit resolves the merge conflict. The commit message literally says: “fix merge conflict“ ;-)

Spacechild1 commented:
@sonoro1234 did you pull latest develop first?

sonoro1234 commented:
Yes, develop is up to date. The problem appears after doing check-out to Spacechild1/supernova_fixes and then pull from Spacechild1/supernova_fixes

Spacechild1 commented:
Dunno. Delete your local branch of supernova_fixes and fetch again.

Spacechild1 commented:
The problem is that SuperCollider's policy is to fix merge conflicts by rebasing instead of merging the base branch. This often (always?) implies a force push, so other people's copies will be out of sync.

Spacechild1 commented:
@brianlheim Can I add this to the 3.10.4 milestone?

brianlheim commented:
@Spacechild1 that is a question for @snappizz

dyfer commented:
I just wanted to add that this PR seems to fix https://github.com/supercollider/supercollider/issues/3339!

Spacechild1 commented:
@dyfer ah yes, because the send_trigger function forgot to lock the memory pool

Can someone pleeeeease merge this PR? I've been waiting half a year now...

brianlheim commented:
@Spacechild1 it needs to be reviewed before it can be merged. I will review it when I have time. It is on my todo list.

Spacechild1 commented:
@brianlheim thanks! sorry for being pushy, but the bug fixes in this PR are rather trivial and it shouldn't take more than a couple of minutes to review them, so I don't see why I should have to wait half a year in an otherwise very active project. Again, the code works and has been tested by several people.

brianlheim commented:
i don't get paid to work on this project. i'll review it when i have time.

Spacechild1 commented:
I don't get paid either. Usually, I would be the one to review and merge Supernova code, but by politics I can't review my own code. This rule is generally great - as long as there's more than one developer with the necessary domain knowledge. So I'm really stuck here. If you (or anyone else) doesn't have time to get familiar with Supernova, that's totally ok, but can't you just make an exception and just merge this? A situation like this is a bit frustrating and makes me think twice before doing any more work on Supernova, please understand.

joshpar commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -203,16 +204,16 @@ template <typename T> static inline void consume(T&& object) {
-    char buffer[128];
-    osc::OutboundPacketStream p(buffer, 128);
+    char buffer[1024];
+    osc::OutboundPacketStream p(buffer, 1024);
```
I would rather see this coded into a single variable that describes what this size means, and use that variable in place of coding 1024 into these places directly. That way, in the future if there is a change like this, it can be done in one location.

Spacechild1 replied:
I see what you mean. 1024 is just one of several OSC message buffer sizes that Supernova uses throughout the code. I've spotted 128, 1024, 2048, 4096 and some variable length buffers. Of course, it would be nice to have const variables instead of magic numbers, but I fear this is a seperate project...

joshpar commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -3626,6 +3648,40 @@ void sc_osc_handler::do_asynchronous_command(World* world, void* replyAddr, cons
+        handle_message_to_RT<true>(world, msg);
+    else
+        handle_message_to_RT<false>(world, msg);
+}
```
could this be simplified to 
`handle_message_to_RT<world->mRealTime>(world, msg);`
?

Spacechild1 replied:
no, because `mRealTime` is a runtime variable but template arguments must be compile time constants.

joshpar commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -3626,6 +3648,40 @@ void sc_osc_handler::do_asynchronous_command(World* world, void* replyAddr, cons
+        handle_message_from_RT<true>(world, msg);
+    else
+        handle_message_from_RT<false>(world, msg);
+}
```
could this be simplified to
`handle_message_from_RT <world->mRealTime>(world, msg);`
?

Spacechild1 replied:
no, because `mRealTime` is a runtime variable but template arguments must be compile time constants.

joshpar commented on server/supernova/sc/sc_ugen_factory.cpp:
```diff
@@ -67,6 +67,7 @@ Unit* sc_ugen_def::construct(sc_synthdef::unit_spec_t const& unit_spec, sc_synth
 
     /* initialize members from synth */
     unit->mParent = static_cast<Graph*>(s);
+    unit->mParentIndex = index;
```
I don't see a reference to this member anywhere else in this PR, so for my own curiosity - is this a new member of Unit?

Spacechild1 replied:
`mParentIndex` is the index of the Ugen within the Synth. It has always been a member of Unit but Supernova simply forgot about it.

joshpar replied:
ok - I was concerned with breaking the UGen interface. Thanks!

Spacechild1 commented:
@joshpar thanks for jumping in! I've answered all your comments. So far there are no updates from my side.

joshpar commented:
@Spacechild1 - since we were discussing the avoidance of magic numbers - can the number of bytes to be skipped be replaced with a size check on something? Or perhaps a comment with why it is sometimes 4, sometimes 8?
`        handle_b_gen<realtime>(message, msg_size, endpoint, 8);
`
`        handle_u_cmd(message, msg_size, 4);
`

Spacechild1 commented:
If the address is an integer, we skip 4 bytes; if the address is a string, we skip the size of OSC string (which is the number of characters + null byte, rounded up to a multiple of 4)

For the integer case I could write `sizeof(int32)` instead of 4, but for the string case a comment would probably make more sense. What do you think?

joshpar commented:
> If the address is an integer, we skip 4 bytes; if the address is a string, we skip the size of OSC string (which is the number of characters + null byte, rounded up to a multiple of 4)
> 
> For the integer case I could write `sizeof(int32)` instead of 4, but for the string case a comment would probably make more sense. What do you think?

yes please - I think so. Would you like me to comment in the PR directly? Or are you ok making the change?

Spacechild1 commented:
I can do it.

Spacechild1 commented:
the macOS CI pipeline failed because of some CMake configuration error...

joshpar commented:
@Spacechild1 - I may have time this evening to dig into the cmake problem

dyfer commented:
> @Spacechild1 - I may have time this evening to dig into the cmake problem

Hey @joshpar could you try restarting Travis build? that sometimes help with random errors (?) and I don't seem to have access to do that

joshpar commented:
Qt is failing to build. It may have to do with the version of Xcode. I'll try to dig into it more later. Sorry about the delay @Spacechild1 and @dyfer

brianlheim commented:
> I don't get paid either.

then show some empathy, _please_. i am trying my best, just like the other maintainers of this project. i don't like that your PR is delayed either. if you look at the open PRs you will see that there are many in the same situation as yours. you will see that this PR involving supernova has very little to do with the delay; in fact, it seems to have received much more attention than most of the others.

brianlheim commented:
> the macOS CI pipeline failed because of some CMake configuration error...

I've pushed a fix for that at #4622 , looks like pinning our CI to an old version of xcode finally caught up with us.

Spacechild1 commented:
@brianlheim Sorry, I don't meant to put any pressure on you personally. Maybe I got just impatient because my other Supernova bug fixes were merged relatively quickly and this one takes so long.

> in fact, it seems to have received much more attention than most of the others.

I hope it does, because it fixes several Server crasher bugs :-) Anyway, thanks for taking your time. I'll stop complaining now.

brianlheim commented on server/supernova/sc/sc_plugin_interface.cpp:
```diff
@@ -563,6 +563,16 @@ int do_asynchronous_command(
     return 0;
 }
 
+bool send_message_from_RT(World* world, struct FifoMsg& msg) {
```
is it OK to use a reference here for C ABI compatibility? or should it be a pointer? (naive question)

Spacechild1 replied:
the plugin API itself defines it like that. in practice, references are implemented as pointers (although the C++ standard doesn't mandate this), so this is safe.

brianlheim commented on server/supernova/sc/sc_osc_handler.hpp:
```diff
@@ -325,6 +327,10 @@ class sc_osc_handler : private detail::network_thread, public sc_notify_observer
 
+    void send_message_from_RT(World* world, FifoMsg& msg);
+
+    void send_message_to_RT(World* world, FifoMsg& msg);
```
these can both be `const` members

Spacechild1 replied:
they could, but for consistency we would have to do this for many other member functions and I think this is out of scope for this PR.

brianlheim replied:
no, you can change it now, some code being const correct is better than none.

brianlheim replied:
this is a typical problem working on this codebase and on legacy and underreviewed codebases in general. many good changes would be held back in the name of "consistency", but modernizing the code in a consistent manner is psychically draining. so, a better strategy is to make incremental improvements and, to the extent that it's possible, ensure that newly added code meets new standards of quality.

Spacechild1 replied:
ok

brianlheim commented:
Please also rebase this against the latest develop branch to pick up the fixes in CI configuration.

Spacechild1 commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -3626,6 +3647,40 @@ void sc_osc_handler::do_asynchronous_command(World* world, void* replyAddr, cons
 }
 
+// called from RT thread, perform in NRT thread, free in RT thread
+template <bool realtime> void handle_message_from_RT(World* world, FifoMsg& msg) {
```
I just tried to be consistent with `do_asynchronous_command`, but you're right that passing the unused parameter is pointless. I just go ahead and remove it from all 3 methods.

Spacechild1 commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -2615,15 +2618,12 @@ template <bool realtime> void handle_b_getn(ReceivedMessage const& msg, endpoint
-template <bool realtime> void handle_b_gen(ReceivedMessage const& msg, size_t msg_size, endpoint_ptr endpoint) {
-    movable_array<char> cmd(msg_size, msg.AddressPattern());
+template <bool realtime>
+void handle_b_gen(ReceivedMessage const& msg, size_t msg_size, endpoint_ptr endpoint, int skip_bytes) {
```
This code has been there before. `TypeTags()` alone is not enough, because we also need to adjust the size, so we would have to do `msg_size - (msg.TypeTags() - msg.AddressPattern())` in all methods which need to instantiate a `sc_msg_iter` - which is redundant because we already know the offset beforehand... Therefore I would prefer to keep it as it is/was.

brianlheim replied:
no, please change it, it's much more expressive. you can write a helper function if you need to.

Spacechild1 replied:
I really don't like to recompute a value we already know. Can't I just add a comment, like `// skip address part`? Or rename `skip_bytes` to a more expressive `address_pattern_size`? I think with the latter it should be really clear what's going on.

brianlheim replied:
no, please change it. as @joshpar  pointed out in a previous review it requires magic numbers and per-usage comments to explain what the current implementation does. that is absolutely not good design. we are talking about subtracting two pointers here, it's really not that much of a sacrifice. if it were, we'd be writing this in C :)

Spacechild1 replied:
on the other hand, it seems that other people also found this part of the code confusing, so I just go ahead and skip the address pattern in the method handler. the runtime cost is negliglible.

Spacechild1 replied:
I've changed my mind in the meantime :-) I saw you comment only now

Spacechild1 replied:
@brianlheim actually, `TypeTags()` returns a pointer to the first type tag, not to the start of the type tag string (which starts with `,`), so I need to write `msg_size - (msg.TypeTags() - msg.AddressPattern() + 1); `. I don't think this is much clearer at all, but I just go ahead and do it.

brianlheim replied:
good catch, thank you for that!

Spacechild1 replied:
found out because my tests failed :-)

Spacechild1 commented on server/supernova/sc/sc_ugen_factory.cpp:
```diff
@@ -67,6 +67,7 @@ Unit* sc_ugen_def::construct(sc_synthdef::unit_spec_t const& unit_spec, sc_synth
 
     /* initialize members from synth */
     unit->mParent = static_cast<Graph*>(s);
+    unit->mParentIndex = index;
```
will do

Spacechild1 commented:
thanks for taking your time I did the requested changes and ran my tests. sorry, rebasing was a mess (again) because the clang-format change happened in the middle of this PR. I accidentally added some config files on the way, but I've removed them.

BTW, is there a reason why clang-format depends on 8.0.0 exactly? I don't manage to install this particular version on my Linux system, so I just gave up and modified clang-format.py to accept 8.0.1 instead (my installed version). if you have a better solution - please tell me :-)

brianlheim commented on server/supernova/sc/sc_osc_handler.cpp:
```diff
@@ -3631,6 +3658,40 @@ void sc_osc_handler::do_asynchronous_command(World* world, void* replyAddr, cons
+    });
+}
+
+void sc_osc_handler::send_message_to_RT(World* world, FifoMsg& msg) const {
```
`world` in `send_message_to_RT` and `send_message_from_RT` (both overloads) should be `const World*`

brianlheim commented on server/supernova/sc/sc_ugen_factory.hpp:
```diff
@@ -67,7 +67,8 @@ class sc_ugen_def : public aligned_class, public named_hash_entry {
         unitcmd_set(unitcmd_set_type::bucket_traits(unitcmd_set_buckets, unitcmd_set_bucket_count)) {}
 
-    Unit* construct(sc_synthdef::unit_spec_t const& unit_spec, sc_synth* s, World* world, linear_allocator& allocator);
+    Unit* construct(sc_synthdef::unit_spec_t const& unit_spec, sc_synth* synth, int synthIndex, World* world,
```
parameter names differ between declaration and definition (synth[Index] vs parent[Index])

brianlheim commented:
> thanks for taking your time! I did the requested changes and ran my tests. sorry, rebasing was a mess (again) because the clang-format change happened in the middle of this PR. I accidentally added some config files on the way, but I've removed them.

oh no! FYI, there's a command in the `clang-format.py` script just for this situation (`clang-format.py rebase`). if I'd remembered your branch pre-dated clang-format, I would have mentioned it!

> BTW, is there a reason why clang-format depends on clang 8.0.0 exactly? I don't manage to install this particular version on my Linux system, so I just gave up and modified clang-format.py to accept clang 8.0.1 instead (my installed version). if you have a better solution - please tell me :-)

i think it was a simple oversight when i initially wrote that script. at that time 8.0.0 was the only 8.y.z release of clang-format and i didn't think there would be another. i'll submit a change soon to accept any 8.y.z version. if there are issues with 8.0.1 (or 8.1.0 if it is released) we can restrict those. probably better to be permissive here.

Spacechild1 commented:
sorry for the oversight - should be fixed now!

snappizz commented:
thanks all for the patience and hard work on this one

sonoro1234 commented:
Congratulations!!!
